subsys/bluetooth/mesh/access.c

```diff
@@ -1388,36 +1387,16 @@ static bool model_has_dst(const struct bt_mesh_model *mod, uint16_t dst, const u
 	return mod->rt->elem_idx == 0;
 }
 
-static const struct bt_mesh_model_op *find_op(const struct bt_mesh_elem *elem,
-					      uint32_t opcode, const struct bt_mesh_model **model)
+static const struct bt_mesh_model_op *find_op_in_list(const struct bt_mesh_model *models,
+						      uint8_t model_count, uint32_t opcode,
+						      const struct bt_mesh_model **model)
 {
-	uint8_t i;
-	uint8_t count;
-	/* This value shall not be used in shipping end products. */
-	uint32_t cid = UINT32_MAX;
-	const struct bt_mesh_model *models;
-
-	/* SIG models cannot contain 3-byte (vendor) OpCodes, and
-	 * vendor models cannot contain SIG (1- or 2-byte) OpCodes, so
-	 * we only need to do the lookup in one of the model lists.
-	 */
-	if (BT_MESH_MODEL_OP_LEN(opcode) < 3) {
-		models = elem->models;
-		count = elem->model_count;
-	} else {
-		models = elem->vnd_models;
-		count = elem->vnd_model_count;
-
-		cid = (uint16_t)(opcode & 0xffff);
-	}
-
-	for (i = 0U; i < count; i++) {
-
+	for (uint8_t i = 0U; i < model_count; i++) {
 		const struct bt_mesh_model_op *op;
 
 		if (IS_ENABLED(CONFIG_BT_MESH_MODEL_VND_MSG_CID_FORCE) &&
-		     cid != UINT32_MAX &&
-		     cid != models[i].vnd.company) {
+		    BT_MESH_MODEL_OP_LEN(opcode) == 3 &&
+		    (uint16_t)(opcode & 0xffff) != models[i].vnd.company) {
```

Comment:
I don't like this tbh. Passing here sig models as `models` but accessing `company` field of vendor models. Even though it is currently safe because [company](https://github.com/zephyrproject-rtos/zephyr/blob/c3190d34aa912d39cd082db822f6ebb332806a2b/include/zephyr/bluetooth/mesh/access.h#L707) and [id](https://github.com/zephyrproject-rtos/zephyr/blob/c3190d34aa912d39cd082db822f6ebb332806a2b/include/zephyr/bluetooth/mesh/access.h#L716) are the same types at the same location, it may change and then this code causes undefined behavior. Neither logically I understand this code.